docs/specification/auth/oauth2.md

```diff
@@ -0,0 +1,285 @@
+---
+title: OAuth 2.0 Authentication
+type: docs
+weight: 51
+---
+
+{{< callout type="info" >}}
+**Protocol Revision**: {{< param protocolRevision >}}
+{{< /callout >}}
+
+The Model Context Protocol (MCP) supports OAuth 2.0 as a standardized authentication method, allowing secure authorization flows between clients and servers. Tokens will be securely communicated as part of the request body.
+
+## Protocol Flow
+
+```mermaid
+sequenceDiagram
+    participant User
+    participant Client as MCP Client
+    participant AuthServer as Authorization<br> MCP Server
+    participant Resource as Resource<br>MCP Server
+
+    User->>Client: Initiate Login
+    Client->>AuthServer: Authorization Request
+    Note over Client,AuthServer: client_id, redirect_uri, scope, state
+    AuthServer->>User: Redirect to Login/Consent/Configuration
+    User->>AuthServer: Authenticate & Authorize
+    AuthServer->>Client: Authorization Code
+    Note over AuthServer,Client: Via redirect_uri with state
+    Client->>AuthServer: Token Request
+    Note over Client,AuthServer: auth code, client_id, client_secret
+    AuthServer->>Client: Access & Refresh Tokens
+    Client->>Resource: Access prompts/resources/tools
+    Resource->>Client: Access roots/sampling
+    
+    Note right of Client: When token expires...
+    Client->>AuthServer: Refresh Token Request
+    AuthServer->>Client: New Access Token
+```
+
+## Capabilities
+
+Clients supporting OAuth 2.0 **MUST** declare it during initialization:
+
+```json
+{
+  "capabilities": {
+    "auth": {
+      "oauth2": true
+    }
+  }
+}
+```
+
+Servers supporting OAuth 2.0 **MUST** include their capabilities:
+
+```json
+{
+  "capabilities": {
+    "auth": {
+      "oauth2": {
+        "token": true,
+        "revoke": true
+      }
+    }
+  }
+}
+```
+
+## Authentication Flows
+
+### Get Access Token Flow
+
+1. Initial Request
+
+When a client without valid credentials attempts to initialize, the server **MUST** respond with an OAuth 2.0 authorization requirement in the form of an error:
+
+```json
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "error": {
+    "code": -32001,
+    "message": "OAuth 2.0 authorization required",
+    "data": {
+      "authRequest": {
+        "authorizationUrl": "<scheme>://<domain>/oauth2",
+        "clientId": "<client_id>",
+        "scope": ["read", "write"],
+        "responseType": "code",
+        "state": "state"
+      }
+    }
+  }
+}
+```
+The authorization URL **SHOULD** prompt the user to configure any credentials as needed.
+
+2. Authorization Code Exchange
+
+After receiving the authorization code, the client **MUST** exchange it for tokens:
```

Comment:
```suggestion
After receiving the authorization code, the client **SHOULD** exchange it for tokens:
```

If the client doesn't do this, they won't be able to initialize, but IMO they should be able to issue another failed request and get another `authRequest` from the server. (Can't think of a practical reason for this, but the sequencing seems fine in the abstract.)